### DISCUSSION TITLE
High Memory Consumption in ETCD During Frequent Updates

### DISCUSSION THREAD

**Discussion by pratiks-alepo** [etcd-io/etcd#15988] [Q&A]
Hello Team,

I am writing to bring your attention to an issue we have been experiencing with ETCD, specifically related to high memory consumption when frequent updates are performed.

We are currently using version 3.4.15 of ETCD, and we have also tested the latest version available, but the behavior remains the same. Whenever we perform frequent updates in ETCD, we notice a significant increase in memory usage, which eventually leads to performance degradation and potential out-of-memory errors.

We have attempted to mitigate this issue by adjusting the snapshot settings as well, but unfortunately, the problem persists. Despite these efforts, the memory consumption continues to rise steadily as the number of updates increases.

We kindly request your assistance in addressing this issue. We would greatly appreciate any guidance, recommendations, or potential solutions that could help alleviate the high memory consumption problem during frequent updates.

Additionally, if there are any known fixes or updates available specifically targeting this memory issue in ETCD, please let us know. We are eager to explore all available options to ensure optimal performance and stability in our environment.

Thank you very much for your attention to this matter. We look forward to your prompt response and any assistance you can provide in resolving this issue.

Best regards,
Pratik

**Comment by jmhbnz** [comment]
Hi @pratiks-alepo - Thanks for raising this discussion, can you please provide a detailed example of what you mean by "frequent updates"?  I.e. what is the operation being performed, including the payload sizes and at what scale?

Additionally can you please provide the configuration file or parameters that you are using for etcd currently?

It would also be helpful context if you could provide some metrics or a summary of what is happening in terms of memory usage.

**Reply by pratiks-alepo** [reply]
Hi @jmhbnz ,

Thank you for your prompt response.The operations we are executing at every 10 seconds are as follows:
`
etcdctl put overload-control/service1/service/status NORMAL
`
`
etcdctl put overload-control/service1/service/status CRITICAL
`

Regarding the ETCD configuration, we have not made any custom changes and are using the default configuration that comes with Docker. We have also experimented with different values of `ETCD_SNAPSHOT_COUNT`, including 1000, 3000, 5000, and 10000. Unfortunately, regardless of the snapshot count, the memory consumption continues to increase steadily over time.

To provide you with a better understanding of the situation, we have attached a memory graph that illustrates the memory usage trend during these frequent update operations.


Best regards,
Pratik

![ETCD_memory](https://github.com/etcd-io/etcd/assets/132875056/07aac98e-a1df-408e-8941-a392f2dfd95e)

**Reply by chaochn47** [reply]
Does the memory usage match with DB size growth?

**Reply by pratiks-alepo** [reply]
Hello @chaochn47 ,

Thank you for your response. We will thoroughly investigate the behavior you mentioned and conduct the necessary tests and observations.

Once we have completed our analysis, we will compile the results and observations for your review. 
We will be in touch with you soon to provide the observed results and any relevant findings.


Best regards,
Pratik

**Reply by fuweid** [reply]
Hi, did you check the container's cgroup memory stat? I mean you check the page cache for the container. Maybe it's related to the db size mentioned by chao.

**Reply by pratiks-alepo** [reply]
Hi @fuweid ,

Thanks for your response.
We are in process to investigate the behavior as mentioned by @chaochn47  and will provide the observed results and any relevant finding as soon as possible.

Best Regards,
Pratik

**Reply by pratiks-alepo** [reply]
Hello  @chaochn47  @fuweid ,

We have completed our analysis regarding the ongoing issue related to ETCD memory growth.Here are our observations:

1. ETCD Key Update: We are updating a single key(as mentioned in earlier post) in ETCD at regular intervals of 10 seconds.

2. Memory Growth: As a result of the frequent key updates, we have observed a steady increase in memory consumption.

3. Database Size: The size of the database is also growing.

To provide more insight into the situation, we have attached the ETCD memory graph and a table illustrating the correlation between the database size and memory consumption.

We welcome any suggestions or insights from the team regarding this matter.


<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/PRATIK~1.SIN/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<link rel=themeData
href="file:///C:/Users/PRATIK~1.SIN/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/PRATIK~1.SIN/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>




Time | ETCD DB-SIZE (in MB) | ETCD Pod Memory (in MB)
-- | -- | --
06:12:25 UTC 2023 | 8.6 | 42
06:17:25 UTC 2023 | 11 | 44
07:27:26 UTC 2023 | 47 | 51
08:02:26 UTC 2023 | 66 | 53
08:32:27 UTC 2023 | 81 | 58
09:02:28 UTC 2023 | 96 | 62
09:32:28 UTC 2023 | 112 | 66
10:02:29 UTC 2023 | 127 | 68
10:32:30 UTC 2023 | 141 | 76




</body>

</html>

![etcd-memory-issue-a1](https://github.com/etcd-io/etcd/assets/132875056/09aa6a7b-203e-4b5b-a62c-012abac3ed97)


Best Regards,
Pratik

**Reply by fuweid** [reply]
Hi @pratiks-alepo according to your last reply, the key `overload-control/service1/service/status` has history versions which seems reasonable to me. You can checkout the https://etcd.io/docs/v3.5/op-guide/maintenance/ to set the `auto-compaction-mode` to delete the history. Hope it can help.

**Reply by pratiks-alepo** [reply]
Hi @fuweid ,

Thanks for your response. We will go through the link and try out some options.
We will get back with our results on this.

Best Regards,
Pratik

**Reply by pratiks-alepo** [reply]
Hello @fuweid,

Thank you for sharing the link. We made some changes to meet our requirements and noticed significant improvement. Here is the current configuration:

- ETCD_AUTO_COMPACTION_MODE: revision
- ETCD_AUTO_COMPACTION_RETENTION: 2
- ETCD_SNAPSHOT_COUNT: 500
- ETCD_MAX_SNAPSHOTS: 2
- ETCD_MAX_WALS: 4

Despite these changes, we are still experiencing a memory leak. Over a span of 7 days, we observed a leak of approximately 30MB, with 20,373,662 update operations performed on the same key. The graph below illustrates this trend.

![image](https://github.com/etcd-io/etcd/assets/132875056/5b36a465-7713-4422-9738-f2a640d122be)



We would appreciate your assistance in identifying any additional changes we can make to address this memory leak issue.

Best regards,
Pratik
